tools/trainerproc/main.c

```diff
@@ -0,0 +1,1965 @@
+    assert(parsed->trainers);
+    for (;;)
+    {
+        while (match_empty_line(p)) {}
+        if (!parse_pragma(p, parsed))
+            break;
+    }
+    for (;;)
+    {
+        if (parsed->trainers_n == trainers_c)
+        {
+            trainers_c *= 2;
+            struct Trainer *trainers_ = realloc(parsed->trainers, sizeof(*parsed->trainers) * trainers_c);
+            assert(trainers_);
+            parsed->trainers = trainers_;
+        }
+        struct Trainer *trainer = &parsed->trainers[parsed->trainers_n];
+        while (match_empty_line(p)) {}
+        if (match_eof(p))
+            break;
+        if (!parse_trainer(p, parsed, trainer))
+        {
+            struct Token t;
+            // Skip to the next trainer.
+            for (;;)
+            {
+                struct Parser p_ = *p;
+                if (match_eof(&p_) || parse_section(&p_, &t))
+                    break;
+                skip_line(p);
+            }
+        }
+        else
+        {
+            parsed->trainers_n++;
+        }
+    }
+}
+
+static void fprint_string(FILE *f, struct String s)
+{
+    fprintf(f, "%.*s", s.string_n, s.string);
+}
+
+static void fprint_bool(FILE *f, bool b)
+{
+    fprintf(f, b ? "TRUE" : "FALSE");
+}
+
+static void fprint_stats(FILE *f, const char *macro, struct Stats stats)
+{
+    fprintf(f, "%s(%d, %d, %d, %d, %d, %d)", macro, stats.hp, stats.attack, stats.defense, stats.speed, stats.special_attack, stats.special_defense);
+}
+
+static bool is_constant(struct String s, const char *prefix)
+{
+    for (int i = 0;; i++)
+    {
+        if (i == s.string_n)
+            return false;
+        else if (prefix[i] == '\0')
+            return s.string[i] == '_';
+        else if (prefix[i] != s.string[i])
+            return false;
+    }
+}
+
+static void fprint_constant(FILE *f, const char *prefix, struct String s)
+{
+    if (!is_constant(s, prefix)) fprintf(f, "%s_", prefix);
+    if (s.string_n > 0)
+    {
+        for (int i = 0; i < s.string_n; i++)
+        {
+            unsigned char c = s.string[i];
+            if (('A' <= c && c <= 'Z') || ('0' <= c && c <= '9'))
+                fputc(c, f);
+            else if ('a' <= c && c <= 'z')
+                fputc(c - 'a' + 'A', f);
+            else
+                fputc('_', f);
+        }
+    }
+    else
+    {
+        fprintf(f, "NONE");
+    }
+}
+
+// This is a really stupid helper for 'fprint_species'.
+static bool is_utf8_character(struct String s, int *i, const unsigned char *utf8)
+{
+    for (int j = *i;; j++, utf8++)
+    {
+        if (*utf8 == '\0')
+        {
+            *i = j - 1; // '- 1' because of the '++' in 'fprint_species'.
+            return true;
+        }
+        else if (j == s.string_n || s.string[j] != *utf8)
+        {
+            return false;
+        }
+    }
+}
+
+static const struct {
+    const char *literal;
+    const char *constant;
+} literal_species[] =
+{
+    { "Calyrex-Ice", "CALYREX_ICE_RIDER" },
```

Comment:
Missing:
- Pikachu-Alola (it's gonna be interpreted as "Alolan" and not "Alola Cap"
- Pikachu-Alola
- Pikachu-Hoenn
- Pikachu-Kalos
- Pikachu-Original
- Pikachu-Partner (it's gonna be interpreted as LGPE's Pikachu instead of "Partner Cap"
- Pikachu-PhD
- Pikachu-Sinnoh
- Pikachu-Starter (LGPE's Pikachu)
- Pikachu-Unova
- Pikachu-World
- Raticate-Alola-Totem
- Marowak-Alola-Totem
- Eevee-Starter
- Unown-Exclamation
- Unown-Question
- Burmy-Sandy
- Burmy-Trash
- Wormadam-Sandy
- Wormadam-Trash
- Mothim-Sandy/Trash (technically, but doesn't matter much within a trainer battle context)
- Shellos-East
- Gastrodon-East
- Basculegion-F
- Darmanitan-Zen
- Darmanitan-Galar-Zen
- Genesect-Burn
- Genesect-Chill
- Genesect-Douse
- Genesect-Shock
- Greninja-Bond
- Scatterbug (technically, but doesn't matter much within a trainer battle context)
- Spewpa (technically, but doesn't matter much within a trainer battle context)
- Vivillon forms
- Flabébé forms
- Floette forms
- Florges forms
- Furfrou forms
- Zygarde forms
- Oricorio-Pa'u
- Rockruff Own Tempo
- Minior forms
- Totem Mimikyu
- Magearna-Original
- Alcremie (^^;)
- Eiscue-Noice
- Morpeko-Hangry
- Zacian-Crowned
- Zamazenta-Crowned
- Oinkologne-F
- Squawkabilly-Blue
- Squawkabilly-White
- Squawkabilly-Yellow

Reply:
Well, we can't have this:
```c
#define SPECIES_PIKACHU_PARTNER                         SPECIES_PIKACHU_PARTNER_CAP
```
Because `SPECIES_PIKACHU_PARTNER` already exists—it's the LGPE Pikachu.

But I've added the rest (that appear in Showdown's team builder for "Anything Goes") to `include/constants/species.h` and the conversion script. There might be other species that ideally would have a different name when converted, but I don't think it's a particularly big problem.